modules/apps/packet-forward-middleware/ibc_middleware_test.go

```diff
@@ -0,0 +1,254 @@
+	pfm := s.pktForwardMiddleware(s.chainA)
+	ack := pfm.OnRecvPacket(ctx, version, channeltypes.Packet{}, relayerAddr)
+	s.Require().False(ack.Success())
+
+	expectedAck := &channeltypes.Acknowledgement{}
+	err := s.chainA.Codec.UnmarshalJSON(ack.Acknowledgement(), expectedAck)
+	s.Require().NoError(err)
+
+	s.Require().Equal("ABCI code: 12: error handling packet: see events for details", expectedAck.GetError())
+}
+
+func (s *PFMTestSuite) TestOnRecvPacket_Nomemo() {
+	s.setupChains()
+
+	ctx := s.chainA.GetContext()
+	version := s.pathAB.EndpointA.GetChannel().Version
+	relayerAddr := s.chainA.SenderAccount.GetAddress()
+	receiverAddr := s.chainB.SenderAccount.GetAddress()
+
+	packet := s.transferPacket(relayerAddr.String(), receiverAddr.String(), s.pathAB, 0, "{}")
+
+	// PacketForwardMiddleware
+	pfm := s.pktForwardMiddleware(s.chainA)
+	ack := pfm.OnRecvPacket(ctx, version, packet, relayerAddr)
+	s.Require().True(ack.Success())
+
+	expectedAck := &channeltypes.Acknowledgement{}
+	err := s.chainA.Codec.UnmarshalJSON(ack.Acknowledgement(), expectedAck)
+	s.Require().NoError(err)
+
+	s.Require().Equal("", expectedAck.GetError())
+	s.Require().ElementsMatch([]byte{1}, expectedAck.GetResult())
+}
+
+func (s *PFMTestSuite) TestOnRecvPacket_InvalidReceiver() {
+	s.setupChains()
+
+	ctx := s.chainA.GetContext()
+	version := s.pathAB.EndpointA.GetChannel().Version
+	relayerAddr := s.chainA.SenderAccount.GetAddress()
+
+	packet := s.transferPacket(relayerAddr.String(), "", s.pathAB, 0, nil)
+
+	// PacketForwardMiddleware
+	pfm := s.pktForwardMiddleware(s.chainA)
+	ack := pfm.OnRecvPacket(ctx, version, packet, relayerAddr)
+	s.Require().False(ack.Success())
+
+	expectedAck := &channeltypes.Acknowledgement{}
+	err := s.chainA.Codec.UnmarshalJSON(ack.Acknowledgement(), expectedAck)
+	s.Require().NoError(err)
+
+	s.Require().Equal("ABCI code: 5: error handling packet: see events for details", expectedAck.GetError())
+	s.Require().Empty(expectedAck.GetResult())
+}
+
+func (s *PFMTestSuite) TestOnRecvPacket_NoForward() {
+	s.setupChains()
+
+	ctx := s.chainA.GetContext()
+	version := s.pathAB.EndpointA.GetChannel().Version
+
+	senderAddr := s.chainA.SenderAccount.GetAddress()
+	receiverAddr := s.chainB.SenderAccount.GetAddress()
+
+	packet := s.transferPacket(senderAddr.String(), receiverAddr.String(), s.pathAB, 0, nil)
+
+	// PacketForwardMiddleware
+	pfm := s.pktForwardMiddleware(s.chainA)
+	ack := pfm.OnRecvPacket(ctx, version, packet, senderAddr)
+	s.Require().True(ack.Success())
+
+	expectedAck := &channeltypes.Acknowledgement{}
+	err := s.chainA.Codec.UnmarshalJSON(ack.Acknowledgement(), expectedAck)
+	s.Require().NoError(err)
+	s.Require().Equal("", expectedAck.GetError())
+
+	s.Require().Equal([]byte{1}, expectedAck.GetResult())
+}
+
+func (s *PFMTestSuite) TestOnRecvPacket_RecvPacketFailed() {
+	s.setupChains()
+
+	transferKeeper := s.chainA.GetSimApp().TransferKeeper
+	ctx := s.chainA.GetContext()
+	// Also can be done if send amount is 0
```

Comment:
```suggestion
```

Not sure what that comment even means, so maybe we just remove it? If you know what it means and it is useful to have, maybe some more context is needed